### PR TITLE
Go CLI Caw implementation

### PR DESCRIPTION
**Pull request by wsuppiger**
Hi Fayez,

Thanks again for looking over this, and really the only file that needs reviewing is `src/go_caw/main.go`.  Everything else is just setup files.  I will add the README for phase 2 later this weekend since it should be pretty straight forward, but other than that this should be the final PR from me.

Best,
William

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -0,0 +1,116 @@
+cloud.google.com/go v0.26.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
```
Comment (floan) [inline go.sum:1]: I believe this file is autogenerated when dependencies are installed using "go mod tidy" so you probably shouldn't push it to GitHub.

File: src/go_caw/main.go [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+}
+
+func main() {
+	flag.Parse()
+
+	// GRPC
+	var conn *grpc.ClientConn
+	conn, err := grpc.Dial("0.0.0.0:50000", grpc.WithInsecure())
+	if err != nil {
+		fmt.Println("sorry, could not connect to Faz")
```
Comment (floan) [inline src/go_caw/main.go:104]: Maybe look into fmt.Errorf() for similar error messages?

File: src/go_caw/main.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+	command := getCommand()
+	event_map := map[EventType]string{
+		kRegisteruser: "registeruser",
+		kCaw:          "caw",
+		kFollow:       "follow",
+		kRead:         "read",
+		kProfile:      "profile",
+	}
+
+	if command == kInvalidCommand {
```
Comment (floan) [inline src/go_caw/main.go:119]: Program should terminate with exit code 1 here as well

File: src/go_caw/main.go [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+			os.Exit(1)
+		} else {
+			read_reply := new(caw.ReadReply)
+			err := ptypes.UnmarshalAny(event_reply.GetPayload(), read_reply)
+			if err != nil {
+				panic(err)
+			}
+			caws := read_reply.Caws
+			// Caws are returned in DFS pattern, so use stack-like structure to format indentation
+			var stack []caw.Caw
```
Comment (floan) [inline src/go_caw/main.go]: Could use a more descriptive name here

**Review by floan — COMMENTED** [review]
Hey William!

Overall great work, have a few tiny comments here and there but in general everything looks good to me.

File: go.sum (hunk 1 above)
Comment (wsuppiger) [reply go.sum:1]: Says on the golang wiki "Ensure your go.sum file is committed along with your go.mod file."

https://github.com/golang/go/wiki/Modules#releasing-modules-all-versions

File: src/go_caw/main.go (hunk 2 above)
Comment (wsuppiger) [reply src/go_caw/main.go:104]: I was using fmt error earlier, but I was running into weird issues for some of my functions.  Let me try it again in my fix commit and see if it works

File: src/go_caw/main.go (hunk 3 above)
Comment (wsuppiger) [reply src/go_caw/main.go:119]: good catch!

File: src/go_caw/main.go (hunk 4 above)
Comment (wsuppiger) [reply src/go_caw/main.go]: I agree, will fix in next commit

**Review by wsuppiger — COMMENTED** [review]
fixing all of these issues

File: src/go_caw/main.go (hunk 2 above)
Comment (wsuppiger) [reply src/go_caw/main.go:104]: yeah I couldn't seem to fix some of the odd formatting, I will keep trying but might just have to leave it

File: go.sum (hunk 1 above)
Comment (floan) [reply go.sum:1]: Really good point, sorry I missed that. This is fine as it is then!

File: src/go_caw/main.go (hunk 2 above)
Comment (floan) [reply src/go_caw/main.go:104]: Yeah, I don't think it's a really major issue. fmt.Println() is fine.

**Review by floan — APPROVED** [review]
Looks good to me, go ahead and merge it in. Congratulations on completing Phase 2!

**Pull request merged**